File: src/main/java/com/leet/easy/Fibonacci.go
package main

/*
509. Fibonacci Number
https://leetcode.com/problems/fibonacci-number/
Easy

The Fibonacci numbers, commonly denoted F(n) form a sequence, called the Fibonacci sequence, such
that each number is the sum of the two preceding ones, starting from 0 and 1. That is,
F(0) = 0, F(1) = 1
F(n) = F(n - 1) + F(n - 2), for n > 1.
Given n, calculate F(n).

Example 1:
Input: n = 2
Output: 1
Explanation: F(2) = F(1) + F(0) = 1 + 0 = 1.

Example 2:
Input: n = 3
Output: 2
Explanation: F(3) = F(2) + F(1) = 1 + 1 = 2.

Example 3:
Input: n = 4
Output: 3
Explanation: F(4) = F(3) + F(2) = 2 + 1 = 3.

Constraints:
    0 <= n <= 30
*/

/* Multiply two matrix */
func mult(arr [2][2]int, base [2][2]int) [2][2]int {
	a, b, c, d := arr[0][0], arr[0][1], arr[1][0], arr[1][1]
	w, x, y, z := base[0][0], base[0][1], base[1][0], base[1][1]
	return [2][2]int{
		{a*w + b*y, a*x + b*z},
		{c*w + d*y, c*x + d*z},
	}
}

/* Fibonacco */
func fib(n int) int {
	if n == 0 {
		return 0
	}

	if n <= 2 {
		return 1
	}

	mat := [2][2]int{{1, 1}, {1, 0}}
	ans := [2][2]int{{1, 1}, {1, 1}}

	n -= 2

	for n != 0 {
		if n%2 == 1 {
			ans = mult(ans, mat)
		}
		mat = mult(mat, mat)
		n /= 2
	}

	return ans[0][0]
}

File: src/main/java/com/leet/easy/Fibonacci.java
package com.leet.easy;

import java.util.stream.IntStream;

public class Fibonacci {
  public int fib_extra_space(int n) {
    int[] arr = new int[31];
    arr[0] = 0;
    arr[1] = 1;
    for (int i = 2; i <= n; i++) {
      arr[i] = arr[i - 2] + arr[i - 1];
    }
    return arr[n];
  }

  public int fibSpaceOptimized(int n) {
    if (n == 0)
      return 0;
    if (n == 1 || n == 2)
      return 1;
    int a = 1, b = 1;
    for (int i = 3; i <= n; i++) {
      int temp = a + b;
      a = b;
      b = temp;
    }
    return b;
  }

  /*
   * a b w x aw + by ax + bz c d y z cw + dy cx + dz
   */
  int[][] multMatrix(int[][] arr, int[][] brr) {
    int a = arr[0][0], b = arr[0][1], c = arr[1][0], d = arr[1][1];
    int w = brr[0][0], x = brr[0][1], y = brr[1][0], z = brr[1][1];
    return new int[][] { new int[] { a * w + b * y, a * x + b * z }, new int[] { c * w + d * y, c * x + d * z } };
  }

  // Log(n)
  public int fibMatrix(int n) {
    if (n == 0)
      return 0;
    if (n == 1)
      return 1;
    int[][] mat = new int[][] { { 1, 1 }, { 1, 0 } };
    int[][] ans = new int[][] { { 1, 1 }, { 1, 1 } };
    n--;
    while (n != 0) {
      if (n % 2 == 1) {
        ans = multMatrix(ans, mat);
      }
      mat = multMatrix(mat, mat);
      n = n >>> 1;
    }
    return ans[0][0];
  }

  public static void main(String[] args) {
    var fib = new Fibonacci();
    IntStream.range(0, 10)
        .forEach(i -> System.out.println(i + " -> " + fib.fib_extra_space(i) + " -> " + fib.fibMatrix(i)));
  }
}
